psi4/driver/mdi_engine.py

```diff
@@ -0,0 +1,417 @@
+            "<FORCES": self.send_forces,
+            ">COORDS": self.recv_coords,
+            ">NLATTICE": self.recv_nlattice,
+            ">CLATTICE": self.recv_clattice,
+            ">LATTICE": self.recv_lattice,
+            ">MASSES": self.recv_masses,
+            "SCF": self.run_scf,
+            "<DIMENSIONS": self.send_dimensions,
+            "<NCOMMANDS": self.send_ncommands,
+            "<COMMANDS": self.send_commands,
+            "<TOTCHARGE": self.send_total_charge,
+            ">TOTCHARGE": self.recv_total_charge,
+            "<ELEC_MULT": self.send_multiplicity,
+            ">ELEC_MULT": self.recv_multiplicity,
+            "EXIT": self.exit
+        }
+
+    def length_conversion(self):
+        """ Obtain the conversion factor between the geometry specification units and bohr
+        """
+        unit_name = self.molecule.units()
+        if unit_name == "Angstrom":
+            unit_conv = MDI_Conversion_Factor("bohr", "angstrom")
+        elif unit_name == "Bohr":
+            unit_conv = 1.0
+        else:
+            raise Exception('Unrecognized unit type: ' + str(unit_name))
+
+        return unit_conv
+
+    # Respond to the <NATOMS command
+    def send_natoms(self):
+        """ Send the number of atoms through MDI
+        """
+        natom = self.molecule.natom()
+        MDI_Send(natom, 1, MDI_INT, self.comm)
+        return natom
+
+    # Respond to the <COORDS command
+    def send_coords(self):
+        """ Send the nuclear coordinates through MDI
+        """
+        coords = self.molecule.geometry().np.ravel()
+        MDI_Send(coords, len(coords), MDI_DOUBLE_NUMPY, self.comm)
+        return coords
+
+    # Respond to the <CHARGES command
+    def send_charges(self):
+        """ Send the nuclear charges through MDI
+        """
+        natom = self.molecule.natom()
+        charges = [self.molecule.charge(iatom) for iatom in range(natom)]
+        MDI_Send(charges, natom, MDI_DOUBLE, self.comm)
+        return charges
+
+    # Respond to the <MASSES command
+    def send_masses(self):
+        """ Send the nuclear masses through MDI
+        """
+        natom = self.molecule.natom()
+        masses = [self.molecule.mass(iatom) for iatom in range(natom)]
+        MDI_Send(masses, natom, MDI_DOUBLE, self.comm)
+        return masses
+
+    # Respond to the <ELEMENTS command
+    def send_elements(self):
+        """ Send the atomic number of each nucleus through MDI
+        """
+        natom = self.molecule.natom()
+        elements = [self.molecule.true_atomic_number(iatom) for iatom in range(natom)]
+        MDI_Send(elements, natom, MDI_INT, self.comm)
+        return elements
+
+    # Respond to the <ENERGY command
+    def send_energy(self):
+        """ Send the total energy through MDI
+        """
+        MDI_Send(self.energy, 1, MDI_DOUBLE, self.comm)
+        return self.energy
+
+    # Respond to the <FORCES command
+    def send_forces(self):
```

Comment:
this and send_energy definitely asymmetric? this runs a gradient QC job, while other returns precomputed? energy.

Reply:
The manual says the following:
```
<ENERGY

The engine sends its most recently calculated energy to the driver. The MD_INIT, SCF, and ATOM_STEP commands can be used to cause the engine to calculate a new energy.

[...]

<FORCES

The engine calculates and sends a set of atomic forces to the driver. These forces include all force contributions, including the force contributions associated with any constraint algorithm (e.g. SHAKE, RATTLE, etc.).
```
```